File: ios/Classes/readers/configurations/EmitterConfigurationReader.swift
//
//  EmitterConfigurationReader.swift
//  snowplow_flutter_tracker
//
//  Created by Matus Tomlein on 03/01/2022.
//

import Foundation
import SnowplowTracker

struct EmitterConfigurationReader: Decodable {
    let bufferOption: String?
    let emitRange: Double?
    let threadPoolSize: Double?
    let byteLimitPost: Double?
    let byteLimitGet: Double?
//    let mockEventStore: Bool?
    
    var bufferOptionType: BufferOption? {
        if let bufferOption = self.bufferOption {
            switch bufferOption {
            case "default":
                return .defaultGroup
            case "heavy":
                return .largeGroup
            case "large":
                return .largeGroup
            default:
                return .single
            }
        }
        return nil
    }
}

extension EmitterConfigurationReader {
    func toConfiguration() -> EmitterConfiguration {
        let configuration = EmitterConfiguration()
        if let bo = self.bufferOptionType { configuration.bufferOption(bo) }
        if let er = self.emitRange { configuration.emitRange(Int(er)) }
        if let tps = self.threadPoolSize { configuration.threadPoolSize(Int(tps)) }
        if let blp = self.byteLimitPost { configuration.byteLimitPost(Int(blp)) }
        if let blg = self.byteLimitGet { configuration.byteLimitGet(Int(blg)) }
        
        return configuration
    }
}

File: android/src/main/kotlin/com/snowplowanalytics/snowplow_flutter_tracker/readers/configurations/EmitterConfigurationReader.kt
package com.snowplowanalytics.snowplow_flutter_tracker.readers.configurations

import com.snowplowanalytics.snowplow.configuration.EmitterConfiguration
import com.snowplowanalytics.snowplow.emitter.BufferOption

class EmitterConfigurationReader(val values: Map<String, Any>) {
    private val valuesDefault = values.withDefault { null }

    val bufferOption: String? by valuesDefault
    val emitRange: Double? by valuesDefault
    val threadPoolSize: Double? by valuesDefault
    val byteLimitPost: Double? by valuesDefault
    val byteLimitGet: Double? by valuesDefault

    fun toConfiguration(): EmitterConfiguration {
        val emitterConfig = EmitterConfiguration()

        bufferOption?.let {
            emitterConfig.bufferOption(when (it) {
                "default" -> BufferOption.DefaultGroup
                "heavy" -> BufferOption.HeavyGroup
                else -> BufferOption.Single
            })
        }
        emitRange?.let { emitterConfig.emitRange(it.toInt()) }
        threadPoolSize?.let { emitterConfig.threadPoolSize(it.toInt()) }
        byteLimitPost?.let { emitterConfig.byteLimitPost(it.toInt()) }
        byteLimitGet?.let { emitterConfig.byteLimitGet(it.toInt()) }

        return emitterConfig
    }
}
